File: r/install_github.R
# Instalação da Biblioteca SDIC-Libraries via GitHub
# Script automatizado para instalar dependências e biblioteca

# Função para instalar bibliotecas se não estiverem presentes
install_if_missing <- function(pkg, repo = NULL) {
  if (!require(pkg, character.only = TRUE, quietly = TRUE)) {
    if (!is.null(repo)) {
      if (!require("remotes", quietly = TRUE)) {
        install.packages("remotes", repos = "https://cloud.r-project.org/")
      }
      remotes::install_github(repo)
    } else {
      install.packages(pkg, repos = "https://cloud.r-project.org/")
    }
  }
}

# Banner de instalação
cat("🚀================================================================🚀\n")
cat("       BIBLIOTECA SDIC - INSTALAÇÃO VIA GITHUB (R)\n")
cat("    Dados de Emprego Brasileiros | Instalação Automática\n")
cat("🚀================================================================🚀\n\n")

# 1. Instalar remotes/devtools se necessário
cat("📦 Instalando ferramentas de desenvolvimento...\n")
install_if_missing("remotes")
install_if_missing("devtools")

# 2. Instalar dependências obrigatórias
cat("\n📋 Instalando dependências da biblioteca...\n")
required_packages <- c(
  "R6",      # Sistema de classes
  "httr2",   # Cliente HTTP moderno
  "jsonlite", # Parser JSON
  "dplyr",   # Manipulação de dados
  "tibble",  # Data frames modernos
  "lubridate", # Manipulação de datas
  "cli",     # Interface linha de comando
  "rlang"    # Ferramentas da linguagem R
)

for (pkg in required_packages) {
  cat("  - Instalando", pkg, "...\n")
  install_if_missing(pkg)
}

# 3. Instalar biblioteca SDIC diretamente do GitHub
cat("\n🚀 Instalando biblioteca SDIC-Libraries do GitHub...\n")
tryCatch({
  remotes::install_github("sdic-org/sdic_libraries", subdir = "r", force = TRUE)
  cat("✅ Biblioteca SDIC-Libraries instalada com sucesso!\n")
}, error = function(e) {
  cat("❌ Erro na instalação:", e$message, "\n")
  cat("💡 Alternativa: Use devtools::install_github('sdic-org/sdic_libraries', subdir='r')\n")
})

# 4. Teste rápido da instalação
cat("\n🧪 Testando instalação...\n")
tryCatch({
  # Tentar carregar o arquivo principal
  source_url <- "https://raw.githubusercontent.com/sdic-org/sdic_libraries/main/r/R/data_access/emprego.R"
  cat("  - Testando carregamento via source...\n")
  cat("  - URL:", source_url, "\n")
  cat("  - Para usar localmente: source('r/R/data_access/emprego.R')\n")
  
  cat("✅ Teste concluído!\n")
}, error = function(e) {
  cat("⚠️ Teste incompleto:", e$message, "\n")
})

# 5. Exemplos de uso
cat("\n📖 EXEMPLOS DE USO:\n")
cat("====================\n")
cat("# Carregar biblioteca via GitHub:\n")
cat("library(sdic.libraries)\n\n")
cat("# Ou carregar arquivo diretamente:\n")
cat("source('https://raw.githubusercontent.com/sdic-org/sdic_libraries/main/r/R/data_access/emprego.R')\n\n")
cat("# Usar a biblioteca:\n")
cat("api <- Emprego$new()\n")
cat("dados <- api$get_saldo_emprego_detalhado('nacional')\n")
cat("length(dados)  # Quantidade de registros\n\n")

cat("🎉 Instalação concluída! Biblioteca pronta para uso.\n")
cat("📖 Documentação: https://github.com/sdic-org/sdic_libraries#readme\n")

File: python/install_github.py
#!/usr/bin/env python3
"""
Instalação da Biblioteca SDIC-Libraries via GitHub (Python)
Script automatizado para instalar diretamente do repositório com pip
"""

import subprocess
import sys
import os


def banner():
    print("🚀" + "=" * 64 + "🚀")
    print("       BIBLIOTECA SDIC - INSTALAÇÃO VIA GITHUB (Python)")
    print("    Dados de Emprego Brasileiros | Instalação Automática")
    print("🚀" + "=" * 64 + "🚀")


def run_command(command, description):
    """Executa comando com feedback visual"""
    print(f"\n📦 {description}...")
    try:
        result = subprocess.run(
            command, 
            shell=True, 
            check=True, 
            capture_output=True, 
            text=True
        )
        print(f"✅ {description} - Sucesso!")
        return True
    except subprocess.CalledProcessError as e:
        print(f"❌ {description} - Erro:")
        print(f"   Comando: {command}")
        print(f"   Saída: {e.output}")
        print(f"   Erro: {e.stderr}")
        return False


def check_python_version():
    """Verifica se a versão do Python é compatível"""
    if sys.version_info < (3, 8):
        print(f"❌ Python {sys.version_info.major}.{sys.version_info.minor} não é compatível.")
        print("   Biblioteca requer Python 3.8+")
        return False
    print(f"✅ Python {sys.version_info.major}.{sys.version_info.minor} é compatível")
    return True


def install_from_github():
    """Instala biblioteca diretamente do GitHub via pip"""
    banner()
    
    # Verificar versão Python
    if not check_python_version():
        sys.exit(1)
    
    # URLs de instalação
    github_urls = [
        "git+https://github.com/sdic-org/sdic_libraries.git#subdirectory=python",
        "git+https://github.com/sdic-org/sdic_libraries.git",  # Fallback
    ]
    
    # Tentar atualizar pip primeiro
    run_command(
        f"{sys.executable} -m pip install --upgrade pip",
        "Atualizando pip"
    )
    
    # Tentar instalação do GitHub
    for i, url in enumerate(github_urls, 1):
        print(f"\n🚀 Tentativa {i}: Instalando do GitHub...")
        print(f"   URL: {url}")
        
        success = run_command(
            f"{sys.executable} -m pip install '{url}'",
            f"Instalação da biblioteca (tentativa {i})"
        )
        
        if success:
            print("✅ Instalação do GitHub bem-sucedida!")
            break
    else:
        print("\n❌ Todas as tentativas de instalação falharam.")
        print("💡 ALTERNATIVAS:")
        print("   1. Clone manual:")
        print("      git clone https://github.com/sdic-org/sdic_libraries.git")
        print("      cd sdic_libraries/python")
        print("      pip install .")
        print("   2. Download direto:")
        print("      pip install https://github.com/sdic-org/sdic_libraries/archive/main.zip#subdirectory=python")
        return False
    
    # Teste de importação
    print("\n🧪 Testando instalação...")
    test_success = run_command(
        f"{sys.executable} -c \"from sdic_libraries.dados.emprego import Emprego; print('✅ Import bem-sucedido')\"",
        "Teste de importação"
    )
    
    if test_success:
        print("\n🎉 INSTALAÇÃO CONCLUÍDA COM SUCESSO!")
        print("\n📖 EXEMPLOS DE USO:")
        print("=" * 50)
        print("from sdic_libraries.dados.emprego import Emprego")
        print("")
        print("# Usar a biblioteca:")
        print("with Emprego() as api:")
        print("    dados = api.get_saldo_emprego_detalhado('nacional')")
        print("    print(f'Registros: {len(dados)}')")
        print("")
        print("📖 Documentação: https://github.com/sdic-org/sdic_libraries#readme")
    else:
        print("⚠️ Instalação aparentemente bem-sucedida, mas teste de importação falhou.")
        print("   Pode haver problemas de dependências.")
    
    return test_success


def main():
    """Função principal"""
    try:
        install_from_github()
    except KeyboardInterrupt:
        print("\n❌ Instalação cancelada pelo usuário.")
        sys.exit(1)
    except Exception as e:
        print(f"\n❌ Erro inesperado: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()